program-runtime/src/loaded_programs.rs

```diff
@@ -228,16 +228,22 @@ impl LoadedProgram {
         elf_bytes: &[u8],
         account_size: usize,
         metrics: &mut LoadProgramMetrics,
+        reloading: bool,
```

Comment:
I'm not very familiar with this code, but this PR lgtm. 

That said, the API now exposes a potential footgun: you get the
reloading argument wrong and we end up executing untrusted code, which
potentially has catastrophic consequences. 

We don't do what I'm about to suggest a lot (sadly), so I'm fine if you want to
merge this as is. However I think we should get into the habit of _not_ having
footguns at the type system level. In this case, I think a safer and rustier API
could be:

```
fn new(
	...leave it as is, no reloading argument
)

/// Reloads an user program, *without* running the verifier.
///
/// This method is unsafe since it assumes that the program has already been verified. Should
/// only be called when ...
unsafe fn reloaded(
	...
)
```

The methods could call a new *private* `new_internal` constructor which does
what `new()` does now in this PR.

With this API, all but one call sites wouldn't have to worry about potentially
executing unverified code. The (future) call site that will reload, will have a
safety comment like: 

```
// Safety: this is safe because ...
unsafe { LoadedProgram::reload(...) };
```

I realize that in this context this might seem like somewhat of a pedantic/paranoid
remark! But I do wish we did this *everywhere* in the validator, and we gotta
start somewhere :P

Reply:
I like the idea of keeping the reload as a separate API that requires an explicit call to it. I'll update the PR.